main/autogen/autoloader.cc

```diff
@@ -0,0 +1,350 @@
+                                              std::unique_ptr<DefTree> &root, ParsedFile &pf) {
+    ENFORCE(root->root());
+    if (!alConfig.includePath(pf.path)) {
+        return;
+    }
+    for (auto &def : pf.defs) {
+        if (def.id.id() == 0) {
+            continue;
+        }
+        addSingleDef(ctx, alConfig, root, NamedDefinition::fromDef(ctx, pf, def.id));
+    }
+}
+
+void DefTreeBuilder::addSingleDef(core::Context, const AutoloaderConfig &alCfg, std::unique_ptr<DefTree> &root,
+                                  NamedDefinition ndef) {
+    if (!alCfg.include(ndef)) {
+        return;
+    }
+    DefTree *node = root.get();
+    for (const auto &part : ndef.nameParts) {
+        auto &child = node->children[part];
+        if (!child) {
+            child = make_unique<DefTree>();
+            child->nameParts = node->nameParts;
+            child->nameParts.emplace_back(part);
+        }
+        node = child.get();
+    }
+    if (ndef.def.defines_behavior) {
+        node->namedDefs.emplace_back(move(ndef));
+    } else {
+        node->nonBehaviorDefs.emplace_back(move(ndef));
+    }
+}
+
+DefTree DefTreeBuilder::merge(DefTree lhs, DefTree rhs) {
+    ENFORCE(lhs.nameParts == rhs.nameParts, "Name mismatch for DefTreeBuilder::merge");
+    lhs.namedDefs.insert(lhs.namedDefs.end(), make_move_iterator(rhs.namedDefs.begin()),
+                         make_move_iterator(rhs.namedDefs.end()));
+    lhs.nonBehaviorDefs.insert(lhs.nonBehaviorDefs.end(), make_move_iterator(rhs.nonBehaviorDefs.begin()),
+                               make_move_iterator(rhs.nonBehaviorDefs.end()));
+    for (auto &[rname, rchild] : rhs.children) {
+        auto lchild = lhs.children.find(rname);
+        if (lchild == lhs.children.end()) {
+            lhs.children[rname] = move(rchild);
+        } else {
+            lhs.children[rname] = make_unique<DefTree>(merge(move(*lchild->second), move(*rchild)));
+        }
+    }
+    return lhs;
+}
+
+void DefTreeBuilder::collapseSameFileDefs(core::Context ctx, const AutoloaderConfig &alCfg, DefTree &root) {
+    core::FileRef definingFile;
+    if (!root.namedDefs.empty()) {
+        definingFile = root.file();
+    }
+    if (!alCfg.sameFileCollapsable(root.nameParts)) {
+        return;
+    }
+
+    for (auto it = root.children.begin(); it != root.children.end(); ++it) {
+        auto &child = it->second;
+        if (child->hasDifferentFile(definingFile)) {
+            collapseSameFileDefs(ctx, alCfg, *child);
+        } else {
+            root.children.erase(it);
+        }
+    }
+}
+
+void AutoloadWriter::writeAutoloads(core::Context ctx, const AutoloaderConfig &alCfg, const std::string &path,
+                                    const DefTree &root) {
+    UnorderedSet<string> toDelete; // Remove from this set as we write files
+    if (FileOps::exists(path)) {
+        vector<string> existingFiles = FileOps::listFilesInDir(path, {".rb"}, true, {}, {});
```

Comment:
This would leave around empty dirs or non `.rb` files. I think that's ok given the way it'd be used, thoughts?